Add tls and password change instructions (#10)
Azure/kubeflow-aks: Deploy-with-tls.md

> # Deploy Kubflow with TLS and Stronger Password

Reviewer: ```suggestion
# Deploy Kubeflow with Password, Ingress and TLS
```

Use the [Azure CLI](https://docs.microsoft.com/cli/azure/install-azure-cli) and [Bicep](https://docs.microsoft.com/azure/azure-resource-manager/bicep/overview) templates to deploy the infrastructure for your application. We will be using the [AKS construction](https://github.com/Azure/AKS-Construction) project to rapidly deploy the required Azure resources. The project allows users the flexibility to tweak their AKS environment however they want. Please check out the [AKS construction helper](https://azure.github.io/AKS-Construction/) for more details about AKS construction.

> You can also try out the automated option using [Mage](https://magefile.org/) build tool at the [Azure open source labs](https://github.com/Azure-Samples/azure-opensource-labs/tree/main/cloud-native/aks-kubeflow#kubeflow-on-azure-kubernetes-service-aks).

Reviewer: ```suggestion
You can also try out the automated option using [Mage](https://magefile.org/) build tool at the [Azure Open Source Labs](https://github.com/Azure-Samples/azure-opensource-labs/tree/main/cloud-native/aks-kubeflow#kubeflow-on-azure-kubernetes-service-aks).
```

```bash
az aks get-credentials --resource-group $RGNAME \
  --name $AKSCLUSTER

kubelogin convert-kubeconfig -l azurecli
```

## Install kustomize

Next install kustomize using the [installation instructions](https://kubectl.docs.kubernetes.io/installation/kustomize/) appropriate for your computer.

> :bulb: In order to use the `kustomize` command below to deploy Kubeflow, you must use [Kustomize v3.2.0](https://github.com/kubernetes-sigs/kustomize/releases/tag/v3.2.0). More info [here](https://github.com/kubeflow/manifests#prerequisites).

## Install kubelogin with TLS and Unique Password
It is assumed you have completed the steps in the main [README](./README.md) file up to point where you deployed installed Kustomize. Follow the steps below to deploy kubeflow and provide access to it with TLS. Please note that a self-signed certificate is used for demonstration purposes. Do not use self signed certs for production workloads. You can easily swap this self-signed cert with your CA certificate for your usecase.

>     > :warning:For this deployment, we will be using a simple method for authenticating to Kubeflow. For more advanced usecases, please configure your deployment to use Azure AD.

Reviewer: ```suggestion
    > :warning: For this deployment, we will be using a simple method for authenticating to Kubeflow. For more advanced usecases, please configure your deployment to use Azure AD.
```

## Install kubelogin with TLS and Unique Password

> It is assumed you have completed the steps in the main [README](./README.md) file up to point where you deployed installed Kustomize. Follow the steps below to deploy kubeflow and provide access to it with TLS. Please note that a self-signed certificate is used for demonstration purposes. Do not use self signed certs for production workloads. You can easily swap this self-signed cert with your CA certificate for your usecase.

Reviewer: Ideally we should combine these instructions with a _copy_ of the AKS deployment instructions from README.md, so that the manifests are overwritten, and deployed in one shot, to avoid exposing the cluster with default credentials.

    > :warning:For this deployment, we will be using a simple method for authenticating to Kubeflow. For more advanced usecases, please configure your deployment to use Azure AD.

> 1. The first step is to generate a new Hash/Password combination using bycrypt. There are many ways of doing this, eg by generating it [using python](https://www.geeksforgeeks.org/hashing-passwords-in-python-with-bcrypt/). For simplicity we will be using coderstool's [Bycrypt Hash Generator](https://www.coderstool.com/bcrypt-hash-generator) for testing purposes. Do not do this for production workloads. In the plane text field, enter a password for your first user, then click on the "Generate Hash" button. You can generate multiple if you have multiple users.

Reviewer: ```suggestion
1. The first step is to generate a new Hash/Password combination using bycrypt. There are many ways of doing this, eg by generating it [using python](https://github.com/kubeflow/manifests/blob/master/README.md#change-default-user-password). For simplicity we will be using coderstool's [Bycrypt Hash Generator](https://www.coderstool.com/bcrypt-hash-generator) for testing purposes. Do not do this for production workloads. In the plain text field, enter a password for your first user, then click on the "Generate Hash" button. You can generate multiple if you have multiple users.
```
We could link to the python section in the official kubeflow docs here.

1. Update your auth.md file with the new email address and password (plain text password not hash) or store the secrets in a more secure way
1. Copy the contents of this newly updated manifests folder to the kubeflow manifests folder. This will update the files so the deployment includes your config changes.
    ```bash
    cp -r tls-manifest/manifests/ manifests/
    ```
1. cd to the manifests folder and install kubeflow
    ```bash
    cd manifests
    ```
    Install all of the components via a single command
    
    ```bash
    while ! kustomize build example | kubectl apply -f -; do echo "Retrying to apply resources"; sleep 10; done
    ```  
1. Once the command has completed, check the pods are ready
    
    ```bash
    kubectl get pods -n cert-manager
    kubectl get pods -n istio-system
    kubectl get pods -n auth
    kubectl get pods -n knative-eventing
    kubectl get pods -n knative-serving
    kubectl get pods -n kubeflow
    kubectl get pods -n kubeflow-user-example-com
    ```

> 1. Restart dex to ensure everything is working as expected

Reviewer: We should stress that it is critical to restart dex to use the new password configuration. Otherwise the default password `12341234` will be used from the time the Service is exposed via `LoadBalancer` until the time this command is run.

    ```bash
    kubectl rollout restart deployment dex -n auth
    ```
1. Configure TLS. Start by getting IP address of istio gateway
    ```bash
    IP=$(kubectl -n istio-system get service istio-ingressgateway --output jsonpath={.status.loadBalancer.ingress[0].ip})
    ```

>     Replace the IP address in the tls-manifest\certificate.yaml file with the IP address of the istio gateway using the sed command below 

Reviewer: ```suggestion
    Replace the IP address in the tls-manifest/certificate.yaml file with the IP address of the istio gateway using the sed command below 
```

    > :warning: If you are using a mac you will need to change the command to `sed -i '' "s/192.168.0.5/$IP/" tls-manifest/certificate.yaml `. 
    
    > :bulb: If these sed commands don't work for any reason or if you don't have sed installed, you will need to update these files manually by replacing the placeholders in the files mentioned below.
    ```bash
    cd ..
    sed -i  "s/192.168.0.5/$IP/" tls-manifest/certificate.yaml 
    ```

> 1. Please note that instead of providing the IP address like we did above, you could give the LB an Azure sub-domain and use that too. Deploy the certificate manifest file.

Reviewer: ```suggestion
1. Please note that instead of providing the IP address like we did above, you could give the LoadBalancer an Azure sub-domain (via the annotation in tls-manifest/manifests/common/istio-1-14/istio-install/base/patches/service.yaml ) and use that too. Deploy the certificate manifest file.
```
